### DISCUSSION TITLE
Node or Flow based Programming UI Component

### DISCUSSION THREAD

**Discussion by PranitModak** [zauberzeug/nicegui#1113] [Ideas / Feature Requests]
### NODE FLOW UI Component

There is vue library/addon which helps creating node based applications which are used some of the top trending web apps today.

_**Vue-Flow Github**_ :  https://github.com/bcakmakoglu/vue-flow
_**Vue-Flow Home**_ :  https://vueflow.dev

It would be really great if we can integrate this in NiceGUI.

Also this **library offers each node execution** which if we can **connect it to backed python for function block execution**, then we get a really good node or flow based programming sandbox builder in place.

@rodja 

![image](https://github.com/zauberzeug/nicegui/assets/43670565/f8f3c51b-dfed-4fb2-bf45-020f1c11efc1)

**Comment by falkoschindler** [comment]
This sounds interesting, but rather challenging. Any help towards implementing such an element is greatly appreciated!

**Comment by rodja** [comment]
pranitrock over at Discord [seems to have a first proof-of-concept](https://discord.com/channels/1089836369431498784/1089836370043883654/1125878603884810360).

**Comment by Stefan2016** [comment]
I found this discussion, when I was searching for a flowchart library to be used within nicegui. Any updates regarding this?
I am just looking to solve this use case: users can draw custom flow charts within a nicegui app and those flow charts will be saved and possibly exported

**Comment by PranitModak** [comment]
Sorry guys was a bit busy last few weeks but yes I have the **component ready** and working. I had coded it for version 1.2.x will have to modify it a bit to suit with 1.3.x framework. Shall push my PR by this week.

**Comment by PranitModak** [comment]
@Stefan2016 @rodja @falkoschindler 

Although I am facing few challenges in event handling of the nodes this is what I have achieved so far.

**Adding below element as is simple as this:** *(Flow is the element being included here)*
![image](https://github.com/zauberzeug/nicegui/assets/43670565/ca9a5727-02b2-425b-a46e-c9dc7e18ff55)


**Adding new nodes:**

https://github.com/zauberzeug/nicegui/assets/43670565/48ded166-a8ae-4970-a6df-b7ff9666cec9


**Interacting with node canvas and deleting nodes:**

https://github.com/zauberzeug/nicegui/assets/43670565/9bc62973-5809-400b-a747-f39cb29cd282

**Reply by rodja** [reply]
Looks good so far.

**Reply by Stefan2016** [reply]
@PranitModak that looks cool!  Looking forward to test it on my own.

**Reply by NaMoCv** [reply]
How is the current progress of the work?

**Reply by ghost** [reply]
How is the current progress of the work? Where can I get the code for a study? tks!

**Reply by mingminyu** [reply]
Any progress of this work, if nicegui can intergerate vue flow, it must more powerful in pipeline work

**Comment by lapnd** [comment]
Hi @PranitModak 
Can you share the code?
Thank you!

**Comment by ghost** [comment]
@PranitModak Is this one still actively being worked on? I need it for my project. would be willing to help!

**Reply by PranitModak** [reply]
I have done most of it. Can details after roughly a week along with releasing the code I developed.

**Reply by ghost** [reply]
thank you! looking forward to it.

**Reply by zhuyubei** [reply]
Hi @PranitModak 

How is your progress on it?

**Reply by Stefan2016** [reply]
@PranitModak  I am also looking forward to it :)

**Reply by lyrlyrl** [reply]
Hi @PranitModak looking forward to it.

**Reply by zhuyubei** [reply]
The author just redfined the meaning of "a week" and gave no responses to the community. Now I have to switch to "pyqt" and "Orange3" to finish my project which needs node editor.

**Reply by WorkSync** [reply]
> The author just redfined the meaning of "a week" and gave no responses to the community. Now I have to switch to "pyqt" and "Orange3" to finish my project which needs node editor.

May I ask how you integrated pyqt5 with nickegui? I am currently trying to embed a nickegui in pyqt5

**Reply by zhuyubei** [reply]
> > The author just redfined the meaning of "a week" and gave no responses to the community. Now I have to switch to "pyqt" and "Orange3" to finish my project which needs node editor.
> 
> May I ask how you integrated pyqt5 with nickegui? I am currently trying to embed a nickegui in pyqt5

No.Pure PyQt，without nicegui.

**Reply by WorkSync** [reply]
> > > The author just redfined the meaning of "a week" and gave no responses to the community. Now I have to switch to "pyqt" and "Orange3" to finish my project which needs node editor.
> > 
> > 
> > May I ask how you integrated pyqt5 with nickegui? I am currently trying to embed a nickegui in pyqt5
> 
> No.Pure PyQt，without nicegui.

I want to integrate Vue or React into PyQt to make the front-end more beautiful. Is there any way , Please help me .

**Reply by zhuyubei** [reply]
> > > > The author just redfined the meaning of "a week" and gave no responses to the community. Now I have to switch to "pyqt" and "Orange3" to finish my project which needs node editor.
> > > 
> > > 
> > > May I ask how you integrated pyqt5 with nickegui? I am currently trying to embed a nickegui in pyqt5
> > 
> > 
> > No.Pure PyQt，without nicegui.
> 
> I want to integrate Vue or React into PyQt to make the front-end more beautiful. Is there any way , Please help me .

By be what you need is this: https://github.com/zhiyiYo/PyQt-Fluent-Widgets

**Reply by WorkSync** [reply]
> > > > > The author just redfined the meaning of "a week" and gave no responses to the community. Now I have to switch to "pyqt" and "Orange3" to finish my project which needs node editor.
> > > > 
> > > > 
> > > > May I ask how you integrated pyqt5 with nickegui? I am currently trying to embed a nickegui in pyqt5
> > > 
> > > 
> > > No.Pure PyQt，without nicegui.
> > 
> > 
> > I want to integrate Vue or React into PyQt to make the front-end more beautiful. Is there any way , Please help me .
> 
> By be what you need is this: https://github.com/zhiyiYo/PyQt-Fluent-Widgets

Thank you for recommending this. Actually, I have already tried it and I hope to integrate with the front-end ecosystem to completely liberate the UI.

**Comment by 123swk123** [comment]
here is my initial implementation for one of my project use case, it allows you to define a node and add fields based on user choice.

![node-graphp-editor](https://github.com/zauberzeug/nicegui/assets/903389/2fe7a0ef-c93d-473e-a707-f9912e808ede)

https://github.com/123swk123/nodegraph-editor-nicegui

any comments and suggestions are welcome!

**Reply by ghost** [reply]
looks sleek!

 Ideally you'd want to control the inputs and outputs of each node separately though. typical use case each node is tied to function with some set of inputs and outputs. the node graph lets you string outputs from one node into the inputs of another node. there will be at least one node whose outputs aren't connected to anything, evaluating said node by recursively evaluating its inputs gets you the output of the overall graph.  

for this to be useful it would probably make more sense to set the inputs and output connection points of a node using two separate lists of objects that include a name and type for the connection point

**Comment by atticus-lv** [comment]
I think vue-flow is a good choice https://vueflow.dev/, it basic on react flow, a powerful flow editor that has many feats. It provide a example of using with quasar https://github.com/bcakmakoglu/vue-flow/tree/master/examples/quasar

**Reply by Cane-xx** [reply]
Nice find. I really like the interactivity. Hope someone can port it over to work with NiceGUI.

**Comment by 123swk123** [comment]
for my next project its heavily based on flow graph function so I am planning to integrate with either [JointJS](https://www.jointjs.com) or [vueflow](https://github.com/bcakmakoglu/vue-flow).

currently I am more inclined towards JointJS. probably I will be done with it by end of the year.

**Reply by 123swk123** [reply]
> for my next project its heavily based on flow graph function so I am planning to integrate with either [JointJS](https://www.jointjs.com) or [vueflow](https://github.com/bcakmakoglu/vue-flow).
> 
> currently I am more inclined towards JointJS. probably I will be done with it by end of the year.

Here is the JointJS on NiceGUI. Basic integration is done, and we can incrementally add features in Vue3 and automatically expose them to NiceGUI. Over the next few weeks, I will be slowly working on this and trying to add as many features as possible.

![jointjs-demo](https://github.com/user-attachments/assets/a3979e5e-87b1-4c8a-89e5-2c818daf1af3)

https://github.com/123swk123/jointjs-nicegui

- [x] Basic framework integration of jointjs on nicegui
- [ ] JointJS module exposure to NiceGUI, this gives more transparent control to NiceGui.
  - [ ] dia
  - [ ] graph
  - [ ] paper
  - [ ] shapes

**Comment by 123swk123** [comment]
probably you should try Qt on webgl (Qt GUI on webbrowser) and use Vue to integrate the bussiness logic and server side rending
 
or another option is to use Qt WASM which might be the best solution and less hassle to integrate.
 
~SwK

**Comment by ljilekor** [comment]
Hi everyone,

Just a suggestion regarding nodes for visual programming.
A powerful framework : [LiteGraph.js](https://github.com/jagenjo/litegraph.js)  [demo](https://tamats.com/projects/litegraph/editor/)

An ultimate example of use : [ComfyUI ](https://github.com/comfyanonymous/ComfyUI) 
Really amazing stuff, the most versatile open source ai tool for CG.

Nicegui implementation would be grrreat!

**Comment by ljilekor** [comment]
Python and Qt based [Pyflow ](https://github.com/wonderworks-software/PyFlow)is inspired on the highly efficient Unreal blueprints

**Comment by Cane-xx** [comment]
I personally only need the most basic flowchart/flow diagram functionality (https://en.wikipedia.org/wiki/Flowchart), not to use the flowchart as a UI for controlling something else. If sky is the limit, then the previous mentioned "Vue Flow" (https://vueflow.dev/) is more aligned with my future needs. But hey I get it, everyone is interested in AI these days...

**Comment by s71m** [comment]
For anyone interested, here’s my implementation of VueFlow for NiceGUI:
[VueFlow for NiceGUI on GitHub](https://github.com/s71m/nicegui_dashboard/blob/master/README.md#vueflow-for-nicegui)

The main challenge was that direct imports for Vue modules are not possible, so I created a UMD bundle instead. Or maybe i dont have right hands to cook it properly

**Reply by falkoschindler** [reply]
Show and tell #4365 🙂

**Reply by zhuyubei** [reply]
> For anyone interested, here’s my implementation of VueFlow for NiceGUI: [VueFlow for NiceGUI on GitHub](https://github.com/s71m/nicegui_dashboard/blob/master/README.md#vueflow-for-nicegui)
> 
> The main challenge was that direct imports for Vue modules are not possible, so I created a UMD bundle instead. Or maybe i dont have right hands to cook it properly

Looks so great! How to support vueflow's minimap function?

**Reply by s71m** [reply]
> > For anyone interested, here’s my implementation of VueFlow for NiceGUI: [VueFlow for NiceGUI on GitHub](https://github.com/s71m/nicegui_dashboard/blob/master/README.md#vueflow-for-nicegui)
> > The main challenge was that direct imports for Vue modules are not possible, so I created a UMD bundle instead. Or maybe i dont have right hands to cook it properly
> 
> Looks so great! How to support vueflow's minimap function?

Since this is a UMD bundle, you need to add the minimap module (as any other module) to the bundle manually.

Update the entry file:
In [docs/create_vueflow_umdjs.md](https://github.com/s71m/nicegui_dashboard/blob/master/docs/create_vueflow_umdjs.md), go to step 3 (src/index.js).
Add the minimap import and export there.

Rebuild the UMD bundle:
After updating index.js, recompile the UMD bundle so that the minimap is included.

Include it in the component:
Finally, update [web/components/vueflow/vueflow.vue](https://github.com/s71m/nicegui_dashboard/blob/c538b7aa8bc1ebca7b6c8f1769be95a412b3208c/web/components/vueflow/vueflow.vue)
to actually import and use the minimap from the new UMD bundle.

**Reply by zhuyubei** [reply]
Thank you for your reply! I'll try it.

**Reply by zhuyubei** [reply]
Hi s71m. I tried to import minimap but did;t find the write way. Because the minimap has to be used in the vueflow in "template", I don't know how to implement this in the nicegui component way. Do you have some ideas?
![image](https://github.com/user-attachments/assets/e78dbbeb-1cde-411c-ae84-bf0e13576036)
